RFCs/0000-expand-new-task.md

```diff
@@ -0,0 +1,158 @@
+
+## Background & Assumptions
+
+- PDK has supported basic Bolt Task generation since version 1.2.0 which was released in October 2017.
+
+- Since that time, Bolt has continued to grow and evolve, and new helper libraries and best practices for writing Bolt Tasks have emerged.
+
+## Motivation & Goals
+
+PDK should provide up-to-date and robust tools for authoring all kinds of Puppet-related content, including Bolt Tasks. The current basic implementation of `pdk new task` can be improved in the following specific ways:
+
+- Users should be able to specify what language they plan to implement a new task in and PDK should generate a language-appropriate scaffold.
+
+- Users should be able to easily convert existing scripts into Bolt Tasks using PDK.
+
+- Users should be able to more easily author "cross-platform" and "shared implementation" tasks.
+
+## Proposed Implementation Design
+
+### Add `--language` option to `pdk new task`
+
+- Extend the current Bolt Task object templates to support language-specific templates.
+
+- Add templates for Bash, PowerShell, Ruby, and Python.
+
+- Ruby and Python templates should have scaffold implementations based on those languages Bolt Task helper libraries. (https://github.com/puppetlabs/puppetlabs-ruby_task_helper and https://github.com/puppetlabs/puppetlabs-python_task_helper)
+
+- Ideally, users could configure a default task language as a module-specific configuration option.
+
+- The new `--language` option will require an argument which is the name of the language-specific task template to be used.
+
+- Acceptable values for the `--language` option will initially be `bash`, `powershell`, `ruby`, and `python`.
+
+- The format of future language option values should be a concise, unique, all lowercase language identifier.
+
+- If an unrecognized language option value is specified, PDK will exit with an error.
+
+- If the `--language` option is specified without an argument, PDK will exit with an error.
+
+#### Examples:
+
+```bash
+$ pdk new task my_default_task
+pdk (INFO): Using default task language: 'bash'
+pdk (INFO): Creating '/Users/jesse/sandbox/pdk/testmod/tasks/my_default_task.sh' from template.
+pdk (INFO): Creating '/Users/jesse/sandbox/pdk/testmod/tasks/my_default_task.json' from template.
+```
+
+```bash
+$ pdk new task --language=ruby my_ruby_task
+pdk (INFO): Creating '/Users/jesse/sandbox/pdk/testmod/tasks/my_ruby_task.rb' from template.
+pdk (INFO): Creating '/Users/jesse/sandbox/pdk/testmod/tasks/my_ruby_task.json' from template.
+```
+
+```bash
+$ pdk new task --language=powershell my_ps_task
+pdk (INFO): Creating '/Users/jesse/sandbox/pdk/testmod/tasks/my_ps_task.ps1' from template.
+pdk (INFO): Creating '/Users/jesse/sandbox/pdk/testmod/tasks/my_ps_task.json' from template.
+```
+
+### Add `--from` option to `pdk new task`
+
+- The new `--from` option will require an argument which is an absolute or relative path to an existing script file.
```

Comment:
It seems like users would want to template both the script and the json file(with `files`) together? I'm not sure how to accomplish that.

Reply:
Can you expand on this? I'm not sure I follow.

The `--from` flag was mostly intended for the "Converting scripts to tasks" workflow where the user would then need to make further modifications to their script to make it a real task. 

But maybe we need to address the "Wrapping an existing script" workflow as well? It should at least be added to the "Future Considerations" I guess.

Reply:
I brought this up again above. Are we expecting that this to be primarily for scripts without arguments? Will in inject some boilerplate param handling commented out at the top or something?

Reply:
I think for now I'm going to drop the `--from` stuff from this RFC, that functionality needs more consideration and I don't want to hold up these other enhancements.